### PR TITLE
*: Update docs with latest Telemeter architecture

### PR DESCRIPTION
**Pull request by saswatamcode**
This PR updates the README with the latest Telemeter architecture and operational paths, along with a diagram.
Feel free to correct me if I am wrong somewhere, as this is not a super familiar path. 🙂 

Rendered here: https://github.com/saswatamcode/telemeter/blob/docs/README.md

Also, will add `mdox` for docs check purposes once in #428 once this is merged!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,74 +1,87 @@
-Telemeter
-=========
+# Telemeter
 
-Telemeter implements a Prometheus federation push client and server
-to allow isolated Prometheus instances that cannot be scraped from a
-central Prometheus to instead perform push federation to a central
-location.
+Telemeter is a set of components used for [OpenShift remote health monitoring](https://docs.openshift.com/container-platform/4.7/support/remote_health_monitoring/about-remote-health-monitoring.html). It allows OpenShift clusters to push telemetry data about clusters to Red Hat, as Prometheus metrics.
```
Comment (PhilipGough) [inline README.md:3]: perhaps change 4.7 -> 4.11 here or `latest` which will keep the link updated.

File: README.md [hunk 2]
```diff
@@ -1,74 +1,87 @@
...
-2. The local client performs cleanup and anonymization and then pushes the metrics to the server.
-3. The server authenticates the client, validates and verifies that the metrics are "safe", and then ensures they have a label uniquely identifying the source client.
-4. The server holds the metrics in a local disk store until scraped.
-5. A centralized Prometheus scrapes each server instance and aggregates all the metrics.
+![Telemeter Architecture](architecture.png)
 
-Since that push is across security boundaries, the server must perform
-authentication, authorization, and data integrity checks as well as being
-resilient to denial of service.
+[telemeter-server](https://github.com/openshift/telemeter/tree/master/cmd/telemeter-server) needs to receive and send metrics across multiple security boundaries, and thus needs to perform serveral authentication, authorization and data integrity checks. It (currently) has two endpoints via which it receives metrics and forwards them to an upstream service as a [Prometheus remote write request](https://github.com/prometheus/prometheus/blob/release-2.38/prompb/remote.proto#L22).
```
Comment (PhilipGough) [inline README.md:7]: typo -> several

File: README.md [hunk 3]
```diff
@@ -1,74 +1,87 @@
...
-instance must ensure that all metrics for a given cluster ID are routed
-to the same instance, otherwise Prometheus will mark those metrics
-series as stale. To do this, the server instances form a cluster using
-a secure gossip transport and build a consistent hash ring so that
-pushed client metrics are routed internally to the same server.
+Telemeter implements a Prometheus federation push client and server to allow isolated Prometheus instances that cannot be scraped from a central Prometheus to instead perform authorized push federation to a central location.
 
-For resiliency, each server instance stores the received metrics on disk
-hashed by cluster ID until they are accessed by a federation endpoint.
+The [telemeter-client](https://github.com/openshift/telemeter/tree/master/cmd/telemeter-client) is deployed via the [OpenShift in-cluster monitoring stack](https://github.com/openshift/cluster-monitoring-operator/blob/master/assets/telemeter-client/deployment.yaml) and performs a certain set of actions via a [`forwarder.Worker`](https://github.com/openshift/telemeter/blob/master/pkg/forwarder/forwarder.go) every 4 minutes and 30 seconds (by default).
```
Comment (JoaoBraveCoding) [inline README.md:13]:
```suggestion
The [telemeter-client](https://github.com/openshift/telemeter/tree/master/cmd/telemeter-client) is deployed via the [OpenShift Cluster Monitoring Operator](https://github.com/openshift/cluster-monitoring-operator/blob/master/assets/telemeter-client/deployment.yaml) and performs a certain set of actions via a [`forwarder.Worker`](https://github.com/openshift/telemeter/blob/master/pkg/forwarder/forwarder.go) every 4 minutes and 30 seconds (by default).
```

File: README.md [hunk 4]
```diff
@@ -1,74 +1,87 @@
...
+2. This handler uses [`tollbooth.NewAuthorizer`](https://github.com/openshift/telemeter/blob/master/pkg/authorize/tollbooth/tollbooth.go) which implements the [`authorize.ClusterAuthorizer`](https://github.com/openshift/telemeter/blob/master/pkg/authorize/cluster.go) interface, to authorize that particular cluster. It uses [`authorize.AgainstEndpoint`](https://github.com/openshift/telemeter/blob/a30ef1e07eb12468cdcbdcf6fde45469a9fa5d60/pkg/authorize/handler.go#L65) to send the cluster ID and token as a POST request to the authorization server (configured via `--authorize`). The authorization server returns a 200 status code, if the cluster is identified correctly.
+3. `tollbooth.AuthorizeCluster` returns a subject which is used as the client identifier in a generated signed JWT which is returned to the telemeter-client, along with any labels.
+
+## /metrics/v1/receive endpoint (receive metrics in prompb.WriteRequest format from any client)
```
Comment (JoaoBraveCoding) [inline README.md:37]: There is no mention but I'm guessing that here we also check if the prometheus is not going over the rate-limiting for the ID of the cluster

Comment (JoaoBraveCoding) [reply README.md:37]: Also, does the whole maintaining timestamp drift also happen on this endpoint?

**Review by JoaoBraveCoding — COMMENTED** [review]
Great job! I also knew very little of this side of things but now it's clearer! Thank you!

File: README.md (hunk 4 above)
Comment (saswatamcode) [reply README.md:37]: So regarding rate limiting requests based on cluster ID, I don't think we do it for metrics/v1/receive currently as per the handler https://github.com/openshift/telemeter/blob/master/cmd/telemeter-server/main.go#L528. This route isn't really used at the moment, except by perhaps Hypershift.

And we do not modify the timestamps in the Prometheus remote write request in any way. We only drop certain timeseries in the request or drop labels. 🙂

Comment (JoaoBraveCoding) [reply README.md:37]: OK perfect! I just mentioned it in case you wanted to mention it on the doc! 👌

**Comment by JoaoBraveCoding** [comment]
Look good to me 💯 Not adding lgmt right away in case @saswatamcode wants the input of more members of the team, otherwise I'm happy to add it

**Review by JoaoBraveCoding — APPROVED** [review]
Look good to me 💯 Not adding lgmt right away in case @saswatamcode wants the input of more members of the team, otherwise, I'm happy to add it

**Review by PhilipGough — APPROVED** [review]
/lgtm

**Pull request merged**